File: 0039_combination_sum/combination_sum.py
from typing import List


def combination_sum(candidates: List[int], target: int) -> List[List[int]]:
    """
    Time:    O(n * log(n) + 2**n)
    Space:   O(n**2)
    """
    res = []
    if not len(candidates):
        return res
    candidates.sort()
    dfs(candidates, [], 0, target, res)
    return res


def dfs(nums: List[int], sub: List[int], index: int, target: int,
        res: List[List[int]]):
    if target == 0:
        res.append(sub)
    for i in range(index, len(nums)):
        if nums[i] > target:
            break
        dfs(nums, sub + [nums[i]], i, target - nums[i], res)

File: 0039_combination_sum/combination_sum.js
/**
 * @param   {number[]}      nums
 * @param   {number[]}      sub
 * @param   {number}        index
 * @param   {number}        target
 * @param   {number[][]}    res
 */
const dfs = (nums, sub, index, target, res) => {
  if (!target) {
    res.push([...sub])
    return
  }
  for (let i = index; i < nums.length; i++) {
    if (nums[i] > target) break
    sub.push(nums[i])
    dfs(nums, sub, i, target - nums[i], res)
    sub.pop()
  }
}

/**
 * Time:    O(n * log(n) + 2**n)
 * Space:   O(n**2)
 * @param   {number[]}      candidates
 * @param   {number}        target
 */
const combinationSum = (candidates, target) => {
  const res = []
  candidates.sort((a, b) => a - b)
  dfs(candidates, [], 0, target, res)
  return res
}
